unit/java_bytecode/java_bytecode_specialize_generics/java_bytecode_specialize_generics.cpp

```diff
@@ -0,0 +1,432 @@
+  "[core][java_bytecode][java_bytecode_specialize_generics]")
+{
+  std::unique_ptr<languaget>java_lang(new_java_bytecode_language());
+
+  // Configure the path loading
+  cmdlinet command_line;
+  command_line.set(
+    "java-cp-include-files",
+    "./java_bytecode/java_bytecode_specialize_generics");
+  config.java.classpath.push_back(
+    "./java_bytecode/java_bytecode_specialize_generics");
+
+  std::istringstream java_code_stream("ignored");
+  null_message_handlert message_handler;
+
+
+  // Configure the language, load the class files
+  java_lang->get_language_options(command_line);
+  java_lang->set_message_handler(message_handler);
+  java_lang->parse(java_code_stream, "generics.class");
+  symbol_tablet new_symbol_table;
+  java_lang->typecheck(new_symbol_table, "");
+  java_lang->final(new_symbol_table);
+
+  generate_java_generic_methodt
+    instantiate_generic_method{message_handler};
+
+
+  GIVEN("Some class files with Generics")
+  {
+    WHEN("Methods with generic signatures")
+    {
+      THEN("Specialize a method with a generic return type")
+      {
+        REQUIRE(
+          new_symbol_table.has_symbol(
+            "java::generics$bound_element.f:()Ljava/lang/Number;"));
+
+        const symbolt &generic_method_symbol=
+          *new_symbol_table.lookup(
+            "java::generics$bound_element.f:()Ljava/lang/Number;");
+
+        const typet &symbol_type=generic_method_symbol.type;
+        REQUIRE(symbol_type.id()==ID_code);
+
+        const code_typet &code=to_code_type(symbol_type);
+        const typet &method_return_type=code.return_type();
+        REQUIRE(is_java_generic_parameter(method_return_type));
+
+        const java_generic_parametert
+          &generic_ret_type=to_java_generic_parameter(method_return_type);
+        const symbol_typet &type_var=generic_ret_type.type_variable();
+        REQUIRE(type_var.get_identifier()==
+                "java::generics$bound_element::NUM");
+
+        const symbolt &specialized_method_symbol=instantiate_generic_method(
+          generic_method_symbol,
+          {
+            { symbol_typet("java::generics$bound_element::NUM"),
+              symbol_typet("java::java.lang.Integer")
+            }
+          },
+          new_symbol_table);
+
+        REQUIRE(
+          new_symbol_table.has_symbol(
+            "java::generics$bound_element.f:()Ljava/lang/Number;"
+              "<java::java.lang.Integer>"));
+        const typet &new_symbol_type=specialized_method_symbol.type;
+        REQUIRE(new_symbol_type.id()==ID_code);
+
+        const code_typet &new_code=to_code_type(new_symbol_type);
+        const typet &new_method_return_type=new_code.return_type();
+        REQUIRE(!is_java_generic_parameter(new_method_return_type));
+
+        REQUIRE(new_method_return_type.subtype()==
+          symbol_typet("java::java.lang.Integer"));
+      }
+
+
+
+      THEN("Specialize a method with a generic parameter types")
+      {
+        REQUIRE(
```

Comment:
I wonder if these "precondition" checks that are really checking the parsing works as you expect it to, should go in there own `THEN` block to make it clear they aren't part of the testing specialisation stuff

Reply:
Yes, that probably makes sense - they act to document what the specialisation stuff is expecting as input, but that is, as you say, just testing the parsing is working.